outline/tun2socks/tunnel_android.go

```diff
@@ -52,3 +54,29 @@ func ConnectShadowsocksTunnel(fd int, client *shadowsocks.Client, isUDPEnabled b
 	go tunnel.ProcessInputPackets(t, tun)
 	return t, nil
 }
+
+// ConnectTunnel reads packets from a TUN device represented by `fd` and routes
+// them to a remote proxy server represented by `configJSON`. This function
+// will also do connectivity tests before starting, so the caller is not
+// required to do any UDP tests.
+//
+// If the function succeeds, it will return a nil error and a Tunnel instance.
+//
+// This function does *not* take ownership of the TUN file descriptor `fd`; the
+// caller is responsible for closing after Tunnel disconnects.
+func ConnectTunnel(configJSON string, fd int) (Tunnel, error) {
+	if len(configJSON) == 0 {
+		return nil, errors.New("tunnel configuration is required")
+	}
+	tunDev, err := tunnel.MakeTunFile(fd)
+	if err != nil {
+		return nil, err
+	}
+	tn, err := newTunnelFromJSON(configJSON, tunDev)
```

Comment:
Remove `tunDev`, and give a better name for the proxy tunnel:
```suggestion
	proxyTun, err := newTunnelFromJSON(configJSON)
```

The initialization that needs `tunDev` should happen below.